lib/stringtheory/src/interning/fixed_size.rs

```diff
@@ -0,0 +1,1002 @@
+unsafe impl Send for InternerState {}
+unsafe impl Sync for InternerState {}
+
+/// A string interner based on a single, fixed-size backing buffer with support for reclamation.
+///
+/// ## Overview
+///
+/// This interner uses a single, fixed-size backing buffer where interned strings are stored contiguously. This provides
+/// bounded memory usage, and the interner will not allocate additional memory for new strings once the buffer is full.
+/// Since interned strings are not likely to need to live for the life of the program, the interner supports
+/// reclamation. Once all references to an interned string have been dropped, the storage for that string is reclaimed
+/// and can be used to hold new strings.
+///
+/// ## Storage layout
+///
+/// The backing buffer stores strings contiguously, with an entry "header" prepended to each string. The header contains
+/// relevant data -- hash of the string, reference count, and length of the string -- needed to work with the entry
+/// either when searching for existing entries or when using the entry itself.
+///
+/// The layout of an entry is as follows:
+///
+/// ┌─────────────────────────────── entry #1 ───────────────────────────────┐ ┌── entry #2 ──┐ ┌── entry .. ──┐
+/// ▼                                                                        ▼ ▼              ▼ ▼              ▼
+/// ┏━━━━━━━━━━━━━┯━━━━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━┓
+/// ┃ string hash │  ref cnt  │ string len │ string data │ alignment padding ┃ ┃    header    ┃ ┃    header    ┃
+/// ┃  (8 bytes)  │ (8 bytes) │ (8 bytes)  │  (N bytes)  │    (variable)     ┃ ┃   & string   ┃ ┃   & string   ┃
+/// ┗━━━━━━━━━━━━━┷━━━━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━┛
+/// ▲                                      ▲                                 ▲
+/// └──────────── `EntryHeader` ───────────┘       alignment matched to ─────┘
+///            (8 byte alignment)                  `EntryHeader` via padding    
+///
+/// The backing buffer is always aligned properly for `EntryHeader`, so that the first entry can be referenced
+/// correctly. However, when appending additional entries to the buffer, we need to ensure that those entries also have
+/// an aligned start for accessing the header. This is complicated due to the variable number of bytes for the string data.
+///
+/// Alignment padding is added to the end of the entry to ensure that when appending the next entry, the start of the
+/// entry is properly aligned for `EntryHeader`. In the worst case, up to 7 bytes could be added (and thus wasted) on
+/// this alignment padding.
+///
+/// ## `InternedString`
+///
+/// The `InternedString` type is a handle to the entry header, and thus the string data, for an interned string. It is
+/// designed to be small -- 8 bytes! -- and cheap to clone, as it contains an atomic reference to the entry header and a
```

Comment:
This claim would be worth putting into a test.

Reply:
Added a test for this in 6a7447b.